tenant/storage_bucket.go

```diff
@@ -0,0 +1,418 @@
+
+func unmarshalBucket(v []byte) (*influxdb.Bucket, error) {
+	u := &influxdb.Bucket{}
+	if err := json.Unmarshal(v, u); err != nil {
+		return nil, ErrCorruptBucket(err)
+	}
+
+	return u, nil
+}
+
+func marshalBucket(u *influxdb.Bucket) ([]byte, error) {
+	v, err := json.Marshal(u)
+	if err != nil {
+		return nil, ErrUnprocessableBucket(err)
+	}
+
+	return v, nil
+}
+
+func (s *Store) GetBucket(ctx context.Context, tx *Tx, id influxdb.ID) (*influxdb.Bucket, error) {
+	encodedID, err := id.Encode()
+	if err != nil {
+		return nil, InvalidOrgIDError(err)
+	}
+
+	b, err := tx.Bucket(bucketBucket)
+	if err != nil {
+		return nil, err
+	}
+
+	v, err := b.Get(encodedID)
+	if kv.IsNotFound(err) {
+		return nil, ErrBucketNotFound
+	}
+
+	if err != nil {
+		return nil, ErrInternalServiceError(err)
+	}
+
+	return unmarshalBucket(v)
+}
+
+func (s *Store) GetBucketByName(ctx context.Context, tx *Tx, orgID influxdb.ID, n string) (*influxdb.Bucket, error) {
+	key, err := bucketIndexKey(orgID, n)
+	if err != nil {
+		return nil, &influxdb.Error{
+			Code: influxdb.EInvalid,
+			Err:  err,
+		}
+	}
+
+	idx, err := tx.Bucket(bucketIndex)
+	if err != nil {
+		return nil, err
+	}
+
+	buf, err := idx.Get(key)
+
+	// allow for hard coded bucket names that dont exist in the system
+	if kv.IsNotFound(err) {
+		switch n {
+		case influxdb.TasksSystemBucketName:
+			return &influxdb.Bucket{
+				ID:              influxdb.TasksSystemBucketID,
+				Type:            influxdb.BucketTypeSystem,
+				Name:            influxdb.TasksSystemBucketName,
+				RetentionPeriod: influxdb.TasksSystemBucketRetention,
+				Description:     "System bucket for task logs",
+				OrgID:           orgID,
+			}, nil
+		case influxdb.MonitoringSystemBucketName:
+			return &influxdb.Bucket{
+				ID:              influxdb.MonitoringSystemBucketID,
+				Type:            influxdb.BucketTypeSystem,
+				Name:            influxdb.MonitoringSystemBucketName,
+				RetentionPeriod: influxdb.MonitoringSystemBucketRetention,
+				Description:     "System bucket for monitoring logs",
+				OrgID:           orgID,
+			}, nil
+		default:
+			return nil, ErrBucketNotFound
+		}
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	var id influxdb.ID
+	if err := id.Decode(buf); err != nil {
+		return nil, &influxdb.Error{
+			Err: err,
+		}
+	}
+	return s.GetBucket(ctx, tx, id)
+}
+
+type BucketFilter struct {
+	Name           *string
+	OrganizationID *influxdb.ID
+}
+
+func (s *Store) ListBuckets(ctx context.Context, tx *Tx, filter BucketFilter, opt ...influxdb.FindOptions) ([]*influxdb.Bucket, error) {
+	// this isn't a list action its a `GetBucketByName`
+	if (filter.OrganizationID != nil && filter.OrganizationID.Valid()) && filter.Name != nil {
+		return nil, invalidBucketListRequest
+	}
+
+	// if we dont have any options it would be irresponsible to just give back all orgs in the system
+	if len(opt) == 0 {
+		opt = append(opt, influxdb.FindOptions{
+			Limit: influxdb.DefaultPageSize,
+		})
+	}
+	o := opt[0]
```

Comment:
I like this idea. One thing I will have to be aware of is that the interface cant change yet because we need this new system to act in pairity with the old kv system. So we cant switch to the `After` style lookups yet. But I do like the pattern of passing find options.

Reply:
Ahh ok cool. I see.